neqo-http3/src/hframe.rs

```diff
@@ -186,124 +201,112 @@ impl HFrameReader {
                 }
             };
 
-            let progress = self.decoder.consume(&mut input);
-            match self.state {
-                HFrameReaderState::BeforeFrame | HFrameReaderState::GetType => match progress {
+            match &mut self.state {
```

Comment:
Since you are here, can you find a way to split this loop out into a new function?